src/bm_sim/lpm_trie.h

```diff
@@ -21,77 +21,79 @@
 #ifndef BM_SIM_LPM_TRIE_H_
 #define BM_SIM_LPM_TRIE_H_
 
-#include <bf_lpm_trie/bf_lpm_trie.h>
+#include <bm/bm_sim/bytecontainer.h>
 
 #include <algorithm>  // for std::swap
-#include <utility>  // for std::swap
+#include <utility>   // for std::swap
+#include <string>
+#include <memory>
+#include <cassert>
+#include <cstdint>
+#include <vector>
 
 namespace bm {
 
-static_assert(sizeof(value_t) == sizeof(uintptr_t),
-              "Invalid type sizes");
+using value_t = std::uintptr_t;
+using byte_t = unsigned char;
+
+static_assert(sizeof(value_t) == sizeof(uintptr_t), "Invalid type sizes");
+
+class Node;
 
 class LPMTrie {
  public:
-  explicit LPMTrie(size_t key_width_bytes)
-    : key_width_bytes(key_width_bytes) {
-    trie = bf_lpm_trie_create(key_width_bytes, true);
-  }
+  explicit LPMTrie(std::size_t key_width_bytes);
 
   /* Copy constructor */
-  LPMTrie(const LPMTrie& other) = delete;
+  LPMTrie(const LPMTrie &other) = delete;
 
   /* Move constructor */
-  LPMTrie(LPMTrie&& other) noexcept
-  : key_width_bytes(other.key_width_bytes), trie(other.trie) {}
+  LPMTrie(LPMTrie &&other) noexcept;
 
-  ~LPMTrie() {
-    bf_lpm_trie_destroy(trie);
-  }
+  ~LPMTrie();
 
   /* Copy assignment operator */
   LPMTrie &operator=(const LPMTrie &other) = delete;
 
   /* Move assignment operator */
-  LPMTrie &operator=(LPMTrie &&other) noexcept {
-    key_width_bytes = other.key_width_bytes;
-    std::swap(trie, other.trie);
-    return *this;
-  }
+  LPMTrie &operator=(LPMTrie &&other) noexcept;
 
+  void insert(const std::string &prefix, int prefix_length, value_t value);
   void insert_prefix(const ByteContainer &prefix, int prefix_length,
-                     uintptr_t value) {
-    bf_lpm_trie_insert(trie, prefix.data(), prefix_length,
-                       static_cast<value_t>(value));
+                     value_t value) {
+    std::string prefix_str(prefix.data(), prefix.size());
+    insert(prefix_str, prefix_length, value);
   }
 
+  bool remove(const std::string &prefix, int prefix_length);
   bool delete_prefix(const ByteContainer &prefix, int prefix_length) {
```

Comment:
Let's move the implementation details of insert_prefix, delete_prefix, has_prefix, retrieve_value in the .cc file